manifest/dynamic_manifest.go

```diff
@@ -31,169 +35,156 @@ type DynamicManifestProvider struct {
 	pubsub           *pubsub.PubSub
 	ec               ec.Backend
 	manifestServerID peer.ID
-	manifestTopic    *pubsub.Topic
 
-	// the lk guards all dynamic manifest-specific fields
-	lk               sync.RWMutex
-	manifest         Manifest
-	onManifestChange OnManifestChange
-	nextManifest     *Manifest
-}
+	runningCtx context.Context
+	errgrp     *errgroup.Group
+	cancel     context.CancelFunc
 
-func NewDynamicManifestProvider(manifest Manifest, pubsub *pubsub.PubSub, ec ec.Backend, manifestServerID peer.ID) ManifestProvider {
-	return &DynamicManifestProvider{
-		manifest:         manifest,
-		pubsub:           pubsub,
-		ec:               ec,
-		manifestServerID: manifestServerID,
-	}
+	initialManifest *Manifest
+	manifestChanges chan *Manifest
 }
 
-func (m *DynamicManifestProvider) Manifest() Manifest {
-	m.lk.RLock()
-	defer m.lk.RUnlock()
-	return m.manifest
+type ManifestUpdateMessage struct {
+	MessageSequence uint64
+	ManifestVersion uint64
+	Manifest        *Manifest
 }
 
-func (m *DynamicManifestProvider) GpbftOptions() []gpbft.Option {
-	return m.manifest.GpbftOptions()
-}
+func (mu ManifestUpdateMessage) toManifest() *Manifest {
+	if mu.Manifest == nil {
+		return nil
+	}
 
-func (m *DynamicManifestProvider) SetManifestChangeCallback(mc OnManifestChange) {
-	m.lk.Lock()
-	defer m.lk.Unlock()
-	m.onManifestChange = mc
+	// When a manifest configuration changes, a new network name is set that depends on the
+	// manifest version of the previous version to avoid overlapping previous configurations.
+	cpy := *mu.Manifest
+	newName := fmt.Sprintf("%s/%d", string(cpy.NetworkName), mu.ManifestVersion)
+	cpy.NetworkName = gpbft.NetworkName(newName)
+	return &cpy
 }
 
-// When a manifest configuration changes, a new network name
-// is set that depends on the manifest version of the previous version to avoid
-// overlapping previous configurations.
-func (m *DynamicManifestProvider) networkNameOnChange() gpbft.NetworkName {
-	return gpbft.NetworkName(string(m.manifest.NetworkName) + "/" + fmt.Sprintf("%d", m.manifest.Sequence))
+func (m ManifestUpdateMessage) Marshal() ([]byte, error) {
```

Comment:
I had to switch to update messages to pull the version out of the manifest (where, IMO, it doesn't really belong). Honestly, I'm not sure if we should even _have_ an explicit version or if we should just use message sequence numbers and version by hashing...